### PR TITLE
MINIFICPP-1997 Fix transiently failing FileSystemRepositoryTests

### PR DESCRIPTION
**Pull request by lordgamez**
https://issues.apache.org/jira/browse/MINIFICPP-1997

------
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Review by martinzink — APPROVED** [review]

**Comment by szaszm** [comment]
Could you explain what's happening here? Why do we have to wait for memory usage to decrease after a small increase? Why do we have to wait more and allow a larger difference? Did minifi become heavier over time, or is there something else at play?

**Comment by lordgamez** [comment]
> Could you explain what's happening here? Why do we have to wait for memory usage to decrease after a small increase? Why do we have to wait more and allow a larger difference? Did minifi become heavier over time, or is there something else at play?

According to @adebreceni the original intent for this test was to check if the allocated memory in the filesystem repository is getting freed after writing the content which is 20MB in this test, so any memory check below 20MB should be sufficient. In very rare cases the memory used by MiNiFi can go above 1MB during the test (which is the original check limit) so it may fail. Usually the memory usage was about 2MB in failure cases, so 5MB should be more than enough to check for. There is also a time limit for this check to allow some time for the agent to release the memory, not just the exact time point after calling the `write()` function.

**Review by adamdebreceni — APPROVED** [review]

File: libminifi/test/unit/FileSystemRepositoryTests.cpp [hunk 1]
```diff
@@ -47,7 +50,6 @@ TEST_CASE("Test Physical memory usage", "[testphysicalmemoryusage]") {
     stream->write(fragment.as_span<const std::byte>());
   }
 
-  const auto end_memory = utils::OsUtils::getCurrentProcessPhysicalMemoryUsage();
-
-  REQUIRE(gsl::narrow<size_t>(end_memory - start_memory) < 1_MB);
+  using org::apache::nifi::minifi::utils::verifyEventHappenedInPollTime;
+  REQUIRE(verifyEventHappenedInPollTime(5s, [&] { return gsl::narrow<size_t>(utils::OsUtils::getCurrentProcessPhysicalMemoryUsage() - start_memory) < 5_MB; }, 100ms));
```
Comment (fgerlits) [inline libminifi/test/unit/FileSystemRepositoryTests.cpp:54]: It's not likely to happen, but I don't think we want to terminate if the current memory usage is lower than `start_memory`.
```suggestion
  CHECK(verifyEventHappenedInPollTime(5s, [&] {
    const auto end_memory = utils::OsUtils::getCurrentProcessPhysicalMemoryUsage();
    REQUIRE(end_memory > 0);
    return end_memory < start_memory + 5_MB;
  }, 100ms));
```

I would also put a `REQUIRE(start_memory > 0)` after line 42.

Comment (lordgamez) [reply libminifi/test/unit/FileSystemRepositoryTests.cpp]: Updated in 45c5e56b0d3804b6fa7f990b4bb3c15666d0de48

Comment (fgerlits) [reply libminifi/test/unit/FileSystemRepositoryTests.cpp]: `gsl::narrow<size_t>(end_memory - start_memory)` still terminates if `end_memory` is less than `start_memory`.  I think `end_memory < start_memory + int64_t{5_MB}` would be better.

Comment (lordgamez) [reply libminifi/test/unit/FileSystemRepositoryTests.cpp]: Oh I see, you are right, fixed in 26e45b1525e76a4ce6d4d2844226868b0c29f068

**Review by fgerlits — APPROVED** [review]

**Review by szaszm — APPROVED** [review]

**Pull request closed without merging**